[Doc]KFP caching (#1875)
kubeflow/website: content/docs/pipelines/caching.md

2.  Follow the [guide to configuring access to Kubernetes
    clusters][kubectl-access]. 

### Disabling caching in your Kubeflow Pipelines deployment:

1. Make sure `mutatingwebhookconfiguration` exists in your cluster:

    ```
    export NAMESPACE=<Namespace where KFP is installed>
    kubectl get mutatingwebhookconfiguration cache-webhook -n ${NAMESPACE}
    ```
2. Change `mutatingwebhookconfiguration` rules:

    ```
    kubectl patch mutatingwebhookconfiguration cache-webhook -n ${NAMESPACE} --type='json' -p='[{"op":"replace", "path": "/webhooks/0/rules/0/operations/0", "value": "DELETE"}]'
    ```

### Enabling caching

1. Make sure `mutatingwebhookconfiguration` exists in your cluster:

    ```
    export NAMESPACE=<Namespace where KFP is installed>
    kubectl get mutatingwebhookconfiguration cache-webhook -n ${NAMESPACE}
    ```
2. Change back `mutatingwebhookconfiguration` rules:

    ```
    kubectl patch mutatingwebhookconfiguration cache-webhook -n ${NAMESPACE} --type='json' -p='[{"op":"replace", "path": "/webhooks/0/rules/0/operations/0", "value": "CREATE"}]'
    ```

## Managing the caching staleness

> 

Reviewer: ```
The cache is enabled by default and if you ever executed same component with the same arguments, any new execution of the component will be skipped and the outputs will be taken from the cache.
For some scenarios, the cached output data of some components might become too stale for use after some time.
To control the maximum staleness of the reused cached data, you can set the step's `max_cache_staleness` parameter.
The `max_cache_staleness` is in [RFC3339 Duration](https://www.ietf.org/rfc/rfc3339.txt) format (so 30 days = "P30D"). 
By default the `max_cache_staleness` is set to infinity so any old cached data will be reused.
```

For a given step, you can set step's `max_cache_staleness` to control the staleness of a target step. The `max_cache_staleness` is in [RFC3339 Duration](https://www.ietf.org/rfc/rfc3339.txt) format(Eg. 30 days = P30D). By default the `max_cache_staleness` will be set to infinity and never gets expired.

Set `max_cache_staleness` to 30 days for a step:

```
def some_pipeline():
      # task is a target step in a pipeline
      task = some_op()
      task.execution_options.caching_strategy.max_cache_staleness = "P30D"
```

> 

Reviewer: ```
Ideally, the component code should be pure and deterministic in the sense that it produces same outputs given same inputs.
If your component is not deterministic (for example, it returns a different random number on every invocation) you might want to disable caching for the tasks created from this component by setting `max_cache_staleness` to 0.
A better solution would be to make the component deterministic. If the component uses random number generation, you can expose the RNG seed as a component input. If the component fetches some changing data you can add a timestamp or date input.
```

+++
title = "Caching"
description = "Getting started with Kubeflow Pipelines step caching"
weight = 50
+++

> Starting from Kubeflow Pipelines 0.4, Kubeflow Pipelines supports step caching capabilities in both standalone deployment and GCP hosted deployment.

Reviewer 2: Nit: Please update instances of 0.4 to 0.4.0.

## Before you start

This guide tells you the basic concepts of Kubeflow Pipelines step caching and how to use it. 
This guide assumes that you already have Kubeflow Pipelines installed or want to use standalone or GCP hosted deployment options in the [Kubeflow Pipelines deployment 
guide](/docs/pipelines/installation/) to deploy Kubeflow Pipelines.

## What is step caching?

> Kubeflow Pipelines caching provides step-level output caching. 

Reviewer 2: Suggested revision for this paragraph:

Kubeflow Pipelines 0.4.0  provides step-level output caching for pipelines built with the Kubeflow Pipelines SDK. 
Step caching is enabled by default. 
With caching enabled, the system can skip steps that have been executed in previous runs,  which saves time and money.
The outputs of cached steps are reused when the cache key of a step matches the cache key of a previously executed step. 
The cache key is based on the component (base image, command-line, code) and arguments passed to the component (values or artifacts).
If the component is exactly the same and the arguments are exactly the same as in some previous execution, then the step can be skipped and the outputs of the previous execution are used. 
Pipeline authors control how long executions are cached by [setting the maximum cache staleness](/docs/pipelines/caching/#managing-caching-staleness).

Pipelines created with the TensorFlow Extended SDK use a different mechanism to cache step execution.

And caching is enabled by default for all pipelines submitted through the KFP backend and UI. 

> The exception is pipelines authored using TFX SDK which has its own caching mechanism. 

Reviewer 2: Is there documentation for the TFX Pipeline caching mechanism? If so, please add a link to it.

Author: Currently not. Thanks!

> The cache key calculation is based on the component (base image, command-line, code), arguments passed to the component (values or artifacts) and any additional customizations. 

Reviewer 2: What are the 'additional customizations' that affect the cache key?

If the component is exactly the same and the arguments are exactly the same as in some previous execution, then the task can be skipped and the outputs of the old step can be used. 
The cache reuse behavior can be controlled and the pipeline author can specify the maximum staleness of the cached data considered for reuse. 
With caching enabled, the system can skip a step that has already been executed which saves time and money. 

## Disabling/enabling caching

> Cache is enabled by default after Kubeflow Pipelines 0.4. 

Reviewer 2: Caching is enabled by default in Kubeflow Pipelines 0.4.0 and higher.

    ```
2. Change back `mutatingwebhookconfiguration` rules:

    ```
    kubectl patch mutatingwebhookconfiguration cache-webhook -n ${NAMESPACE} --type='json' -p='[{"op":"replace", "path": "/webhooks/0/rules/0/operations/0", "value": "CREATE"}]'
    ```

## Managing caching staleness

The cache is enabled by default and if you ever executed same component with the same arguments, any new execution of the component will be skipped and the outputs will be taken from the cache.
For some scenarios, the cached output data of some components might become too stale for use after some time.
To control the maximum staleness of the reused cached data, you can set the step's `max_cache_staleness` parameter.
The `max_cache_staleness` is in [RFC3339 Duration](https://www.ietf.org/rfc/rfc3339.txt) format (so 30 days = "P30D"). 
By default the `max_cache_staleness` is set to infinity so any old cached data will be reused.

Set `max_cache_staleness` to 30 days for a step:

```
def some_pipeline():
      # task is a target step in a pipeline
      task = some_op()
      task.execution_options.caching_strategy.max_cache_staleness = "P30D"
```

> Ideally, the component code should be pure and deterministic in the sense that it produces same outputs given same inputs.

Reviewer 2: Nit: Isn't a function that is pure deterministic by definition?

Ideally, each component should be pure, in that the component deterministically produces the same outputs for a given set of inputs and does not have additional side effects.

+++

> title = "Caching"

Author: I think Caching would be more general.

> title = "Caching"

Reviewer: Another alternative might be Execution caching.

> title = "Caching"

Author: IIRC, we also have the concept of execution in MLMD. Is that the same concept here?

> title = "Caching"

Reviewer: I think the concept is the same.

Component
Task = Component+Arguments (arguments are not fully resolved and might be references to the outputs of other tasks)
Execution = Fully resolved task submitted for execution

> title = "Caching"

Reviewer 2: I would recommend picking a name that describes this feature and updating the title and usage throughout the doc. Step caching and Execution caching both make sense to me.

description = "Getting start with Kubeflow Pipelines step caching"
weight = 50
+++

> Starting from Kubeflow Pipelines 0.4, Kubeflow Pipelines supports the step caching capabilities in both standalone deployment and marketplace deployment.

Author: It will be included only in standalone and marketplace deployment. Not full KF yet.

> Starting from Kubeflow Pipelines 0.4, Kubeflow Pipelines supports the step caching capabilities in both standalone deployment and marketplace deployment.

Reviewer: > Will step caching only work on standalone/marketplace deployments? Or will this work on full KF when KFP 0.4 is part of the KF release? I

KF has their own manifests and release schedule. Caching should be made available in KF's new release (but that will require a bit of work porting and adding the manifests.)

> Starting from Kubeflow Pipelines v0.4, Kubeflow Pipelines supports the step caching capabilities in both standalone deployment and marketplace deployment.

Reviewer 3: Took a look it seems like we never use `v` before the version number.

```suggestion
Starting from Kubeflow Pipelines 0.4, Kubeflow Pipelines supports the step caching capabilities in both standalone deployment and marketplace deployment.
```

## Before you start

> This guide tells you the basic concepts of Kubeflow Pipelines step caching and how to use it. This guide assumes that you already have Kubeflow Pipelines installed or want to use standalone or marketplace deployment options in the [Kubeflow Pipelines deployment 

Author: Thanks! Change "marketplace deployment" to "GCP hosted deployment".

guide](/docs/pipelines/installation/) to deploy Kubeflow Pipelines.

## What is step caching?

> Kubeflow Pipelines step caching provides step-level output caching. For a single pipeline, a step output will be taken from cache if this step template and input does not change. With step caching, the output of a certain step can be shareable and already existing step can prevent re-executing.  

Reviewer 3: ```suggestion
Kubeflow Pipelines step caching provides step-level output caching. A step output will be taken from cache if this step template and input does not change. With step caching, the output of a certain step can be re-used and we can save the efforts of re-executing previous steps.  
```

## Installation

Caching is installed by default after Kubeflow Pipelines v0.4. You can follow the instructions of [Upgrading or reinstalling Kubeflow Pipelines](/docs/pipelines/upgrade) to upgrade or reinstall your Kubeflow Pipelines.

## Control caching behavior

### Configure kubectl to talk to your cluster
See the Google Kubernetes Engine (GKE) guide to [configuring cluster access for kubectl](https://cloud.google.com/kubernetes-engine/docs/how-to/cluster-access-for-kubectl).

### Switch off caching in your Kubeflow Pipelines deployment:

> 1. Make sure mutatingwebhookconfiguration exists in your cluster:

Reviewer 2: Update mutatingwebhookconfiguration to use code font:

`mutatingwebhookconfiguration`

Please update the subsequent instances too.

    ```

>     export NAMESPACE=namespace_kfp_installed

Reviewer 3: Just to prevent blindly copy-n-paste.

```suggestion
    export NAMESPACE=<Namespace where KFP is installed>
```

    kubectl get mutatingwebhookconfiguration cache-webhook -n ${NAMESPACE}
    ```
2. Change mutatingwebhookconfiguration rules:

    ```
    kubectl patch mutatingwebhookconfiguration cache-webhook -n ${NAMESPACE} --type='json' -p='[{"op":"replace", "path": "/webhooks/0/rules/0/operations/0", "value": "DELETE"}]'
    ```

### Switch on caching

1. Make sure mutatingwebhookconfiguration exists in your cluster:

    ```

>     export NAMESPACE=namespace_kfp_installed

Reviewer 3: Same as above.

    kubectl get mutatingwebhookconfiguration cache-webhook -n ${NAMESPACE}
    ```
2. Change back mutatingwebhookconfiguration rules:

    ```
    kubectl patch mutatingwebhookconfiguration cache-webhook -n ${NAMESPACE} --type='json' -p='[{"op":"replace", "path": "/webhooks/0/rules/0/operations/0", "value": "CREATE"}]'
    ```

> ### Per step control: max_cache_staleness

Reviewer 3: ```suggestion
### Per step control: `max_cache_staleness`
```

For a given step, you can set step's max_cache_staleness to control the staleness of a target step. The max_cache_staleness is in [RFC3339 Duration](https://www.ietf.org/rfc/rfc3339.txt) format. Eg. 30 days = P30D.

> Set max_cache_staleness to 30 days for a step:

Reviewer 3: ```suggestion
Set `max_cache_staleness` to 30 days for a step:
```

> Set max_cache_staleness to 30 days for a step:

Reviewer 3: Perhaps also briefly comment on its default staleness?

> For a given step, you can set step's max_cache_staleness to control the staleness of a target step. The max_cache_staleness is in [RFC3339 Duration](https://www.ietf.org/rfc/rfc3339.txt) format(Eg. 30 days = P30D). By default the `max_cache_staleness` will be set to infinity and never gets expired.

Reviewer 2: Update max_cache_staleness, including subsequent instances,  to use code font:

`max_cache_staleness`

Author: Updated max_cache_staleness in context. But in title the code font looks little weird. Should I keep using code font in title?

```
def some_pipeline():
      # task is a target step in a pipeline
      task = some_op()
      task.execution_options.caching_strategy.max_cache_staleness = "P30D"
```

> Setting max_cache_staleness to 0 for a step means this step output will never be taken from cache:

Reviewer 3: Is it possible to use something like `dsl.get_pipeline_conf().add_op_transformer()` to change the `max_cache_staleness` globally? See example [here](https://github.com/kubeflow/pipelines/blob/dbac974c51f9c6c21ce8af495defd55a6fb8a464/samples/core/xgboost_training_cm/xgboost_training_cm.py#L285)

Reviewer: Hmm. We can add this.
Although in practice you can just set it on the first task in the sequence and all subsequent steps will re-run too.

+++
title = "Caching"
description = "Getting started with Kubeflow Pipelines step caching"
weight = 50
+++

Starting from Kubeflow Pipelines 0.4, Kubeflow Pipelines supports step caching capabilities in both standalone deployment and GCP hosted deployment.

## Before you start

This guide tells you the basic concepts of Kubeflow Pipelines step caching and how to use it. This guide assumes that you already have Kubeflow Pipelines installed or want to use standalone or GCP hosted deployment options in the [Kubeflow Pipelines deployment 
guide](/docs/pipelines/installation/) to deploy Kubeflow Pipelines.

## What is step caching?

> Kubeflow Pipelines caching provides step-level output caching on [Kubeflow Pipelines SDK](/docs/pipelines/sdk) only. A step output will be taken from cache if this step template and input does not change. The change includes step name, input artifacts, parameters and underlying logic change. Step output passed in a GCS bucket has not been included in this cache support yet. A cached entry is managed by Kubeflow Pipelines database and will never expire by default. To control caching behavior and change cache entry staleness, please see sections below. With step caching, the output of a certain step can be re-used and we can save the efforts of re-executing previous steps.  

Reviewer: It might be better to have each sentence on its own line so that it's easier to comment.

>Kubeflow Pipelines caching provides step-level output caching on [Kubeflow Pipelines SDK](/docs/pipelines/sdk) only.

I'm not sure I fully understand what's conveyed by this sentence.
Do this mean that "Execution caching is enabled by default for all pipelines submitted through the KFP backend and UX. The exception is pipelines authored using TFX SDK which has its own caching mechanism."

>The change includes step name, input artifacts, parameters and underlying logic change.

The cache key is based on the component (base image, command-line, code), arguments passed to the component (values or artifacts) and any additional customizations. If the component is exactly the same and the arguments are exactly the same as in some previous execution, then the task can be skipped and the outputs of the old execution can be used.

>Step output passed in a GCS bucket has not been included in this cache support yet.

?

>A cached entry is managed by Kubeflow Pipelines database and will never expire by default.

This is a bit vague.

>To control caching behavior and change cache entry staleness, please see sections below.

"The cache reuse behavior can be controlled and the pipeline author can specify the maximum staleness of the cached data considered for reuse."

>With step caching, the output of a certain step can be re-used and we can save the efforts of re-executing previous steps.

"With execution caching enabled, the system can skip a step that has already been executed which saves time and money."
